exporter/loadbalancingexporter/resolver_srv_test.go

```diff
@@ -0,0 +1,463 @@
+// // Copyright The OpenTelemetry Authors
+// // SPDX-License-Identifier: Apache-2.0
+
+package loadbalancingexporter
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestInitialSRVResolution(t *testing.T) {
+	// prepare
+	res, err := newSRVResolver(zap.NewNop(), "_port._proto.service-1.ns.svc.cluster.local", "", 5*time.Second, 1*time.Second)
+	require.NoError(t, err)
+
+	res.resolver = &mockSRVResolver{
+		onLookupIPAddr: func(context.Context, string) ([]net.IPAddr, error) {
+			return []net.IPAddr{
+				{IP: net.IPv4(12, 12, 12, 12)},
+			}, nil
+		},
+		onLookupSRV: func(context.Context, string, string, string) (string, []*net.SRV, error) {
+			return "notapplicable", []*net.SRV{
+				{Target: "pod-0.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-1.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-2.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+			}, nil
+		},
+	}
+
+	// test
+	var resolved []string
+	res.onChange(func(endpoints []string) {
+		resolved = endpoints
+	})
+	require.NoError(t, res.start(context.Background()))
+	defer func() {
+		require.NoError(t, res.shutdown(context.Background()))
+	}()
+
+	// verify
+	assert.Len(t, resolved, 3)
+	for i, value := range []string{
+		"pod-0.service-1.ns.svc.cluster.local",
+		"pod-1.service-1.ns.svc.cluster.local",
+		"pod-2.service-1.ns.svc.cluster.local",
+	} {
+		assert.Equal(t, value, resolved[i])
+	}
+}
+
+func TestInitialSRVResolutionWithPort(t *testing.T) {
+	// prepare
+	res, err := newSRVResolver(zap.NewNop(), "_port._proto.service-1.ns.svc.cluster.local", "12000", 5*time.Second, 1*time.Second)
+	require.NoError(t, err)
+
+	res.resolver = &mockSRVResolver{
+		onLookupIPAddr: func(context.Context, string) ([]net.IPAddr, error) {
+			return []net.IPAddr{
+				{IP: net.IPv4(12, 12, 12, 12)},
+			}, nil
+		},
+		onLookupSRV: func(context.Context, string, string, string) (string, []*net.SRV, error) {
+			return "notapplicable", []*net.SRV{
+				{Target: "pod-0.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-1.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-2.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+			}, nil
+		},
+	}
+
+	// test
+	var resolved []string
+	res.onChange(func(endpoints []string) {
+		resolved = endpoints
+	})
+	require.NoError(t, res.start(context.Background()))
+	defer func() {
+		require.NoError(t, res.shutdown(context.Background()))
+	}()
+
+	// verify
+	assert.Len(t, resolved, 3)
+	for i, value := range []string{
+		"pod-0.service-1.ns.svc.cluster.local:12000",
```

Comment:
This would be really confusing to users: if the srv record shows that the target port is 4317, why is it reaching out to port 12000? It would take people some time to figure out that there's a Port setting on the collector config taking precedence over the SRV record. IMO, the right thing here is to return an error if users set a Port on the collector config, and always use the one from the SRV record.